pkg/register/document.go

```diff
@@ -54,12 +69,120 @@ type RegisterDocument struct {
 	Metadata               Metadata                  `json:"metadata,omitempty"`
 }
 
-// PublicKeyByID get public key by ID (note: excluding authentication keys)
-func (d RegisterDocument) PublicKeyByID(id string) *RegisterPublicKey {
-	for _, v := range d.PublicKeys {
+// PublicKeyByID get public key by ID (note: excluding authentication keys).
+func (document RegisterDocument) PublicKeyByID(id string) *RegisterPublicKey {
+	for _, v := range document.PublicKeys {
 		if v.ID == id {
-			return &v
+			return &RegisterPublicKey{
+				ID:              v.ID,
+				Type:            v.Type,
+				PublicKeyBase58: v.PublicKeyBase58,
+				Revoked:         v.Revoked,
+			}
 		}
 	}
+
 	return nil
 }
+
+// Validate Document validation (correct fields, lengths, types etc NOT CRYPTO)
+func (document RegisterDocument) Validate() []error {
+	var errs []error
+
+	if document.Context != DocumentContext {
+		errs = append(errs, fmt.Errorf("document context must be: '%s'", DocumentContext))
+	}
+
+	if !specVersionExists(document.IoticsSpecVersion) {
+		errs = append(errs, fmt.Errorf("document version should be: '%s'", defaultVersion))
+	}
+
+	err := validation.ValidateIdentifier(document.ID)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	_, err = identity.ParseDidType(document.IoticsDIDType)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	if document.Controller == document.ID {
+		errs = append(errs, errors.New("document controller cannot be self"))
+	}
+
+	if len(document.Metadata.Label) > maxLabelLength {
+		errs = append(errs, fmt.Errorf("metadata label is longer than max %d", maxLabelLength))
+	}
+	if len(document.Metadata.Comment) > maxCommentLength {
+		errs = append(errs, fmt.Errorf("metadata comment is longer than max %d", maxCommentLength))
+	}
+	if len(document.Metadata.URL) > maxURLLength {
+		errs = append(errs, fmt.Errorf("metadata url is longer than max %d", maxURLLength))
+	}
+
+	// Check key/delegation names are unique in the document
+	bufOfNames := ""
+
+	for _, publicKey := range document.PublicKeys {
+		errs = append(errs, validatePublicKey(PublicKeyTypeString, publicKey)...)
+
+		if strings.Contains(bufOfNames, publicKey.ID+"|") {
```

Comment:
To check the unicity we could use a `map[string]struc{}`, the operation on string concatenation and lookup are usually quite expensive. 
(Not the best example case ever but it shows the diff with a poor granularity :
with strings: https://ideone.com/Bt41zr
with map as a set: https://ideone.com/wZVZRl)

Reply:
https://github.com/Iotic-Labs/iotics-identity-go/pull/9